### PR TITLE
[tests] add traffic analysis for Cert_5_3_04_AddressMapCache.py

### PR DESCRIPTION
**Pull request by hjian2017**

### CODE REVIEW

File: tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py [hunk 1]
```diff
@@ -50,25 +52,30 @@ def setUp(self):
...
-        self.nodes[ROUTER1].set_router_selection_jitter(1)
+        self.nodes[DUT_ROUTER1].set_panid(0xface)
+        self.nodes[DUT_ROUTER1].set_mode('rsdn')
+        self.nodes[DUT_ROUTER1].add_whitelist(self.nodes[LEADER].get_addr64())
+        self.nodes[DUT_ROUTER1].add_whitelist(self.nodes[SED1].get_addr64())
+        self.nodes[DUT_ROUTER1].enable_whitelist()
+        self.nodes[DUT_ROUTER1].set_router_selection_jitter(1)
 
         self.nodes[SED1].set_panid(0xface)
         self.nodes[SED1].set_mode('rsn')
```
Comment (librasungirl) [inline tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py]: remove 'r' (Rx-Off-When-Idle) for SED

File: tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py [hunk 2]
```diff
@@ -50,25 +52,30 @@ def setUp(self):
...
-        self.nodes[SED1].add_whitelist(self.nodes[ROUTER1].get_addr64())
+        self.nodes[SED1].add_whitelist(self.nodes[DUT_ROUTER1].get_addr64())
         self.nodes[SED1].enable_whitelist()
 
+        self.nodes[ED1].set_panid(0xface)
```
Comment (librasungirl) [inline tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py]: ED1/2/3/4 have the same configuration. use `for` loop?

File: tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py [hunk 3]
```diff
@@ -84,56 +91,58 @@ def setUp(self):
...
-        time.sleep(5)
-        self.assertEqual(self.nodes[ED3].get_state(), 'child')
+        # This method flushes the message queue so calling this method again will return only the oldly logged messages.
+        dut_messages = self.sniffer.get_messages_sent_by(DUT_ROUTER1)
 
-        self.nodes[ED4].start()
-        time.sleep(5)
-        self.assertEqual(self.nodes[ED4].get_state(), 'child')
+        # 2
+        for i in range(4, 8):
```
Comment (librasungirl) [inline tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py]: avoid to use magic number (4, 8) if possible? e.g. here uses list [ED1, ED2, ED3, ED4] for better readability?

File: tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py [hunk 4]
```diff
@@ -84,56 +91,58 @@ def setUp(self):
...
+            ed_mleid = self.nodes[i].get_ip6_address(config.ADDRESS_TYPE.ML_EID)
+            self.assertTrue(self.nodes[SED1].ping(ed_mleid))
 
-        self.nodes[ED5].start()
+        # verify DUT_ROUTER1 generated an Address Query Request to find each node's RLOC.
+        dut_messages = self.sniffer.get_messages_sent_by(DUT_ROUTER1)
+        msg = dut_messages.next_coap_message('0.02', '/a/aq')
+        command.check_address_query(msg, self.nodes[DUT_ROUTER1], config.REALM_LOCAL_ALL_ROUTERS_ADDRESS)
```
Comment (librasungirl) [inline tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py]: Above check should be in the `for` block and verified per ping request.

File: tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py [hunk 5]
```diff
@@ -84,56 +91,58 @@ def setUp(self):
...
+            self.assertTrue(self.nodes[SED1].ping(ed_mleid))
 
-        self.nodes[ED5].start()
+        # verify DUT_ROUTER1 generated an Address Query Request to find each node's RLOC.
+        dut_messages = self.sniffer.get_messages_sent_by(DUT_ROUTER1)
+        msg = dut_messages.next_coap_message('0.02', '/a/aq')
+        command.check_address_query(msg, self.nodes[DUT_ROUTER1], config.REALM_LOCAL_ALL_ROUTERS_ADDRESS) 
+
+        # 3 & 4
+        # This method flushes the message queue so calling this method again will return only the oldly logged messages.
```
Comment (librasungirl) [inline tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py]: no such word `oldly`

File: tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py [hunk 6]
```diff
@@ -84,56 +91,58 @@ def setUp(self):
...
         time.sleep(5)
-        self.assertEqual(self.nodes[ED5].get_state(), 'child')
+        dut_messages = self.sniffer.get_messages_sent_by(DUT_ROUTER1)
 
         for i in range(4, 8):
```
Comment (librasungirl) [inline tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py]: avoid to use magic number (4, 8) if possible? e.g. here uses list [ED1, ED2, ED3, ED4] for better readability?

File: tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py [hunk 7]
```diff
@@ -84,56 +91,58 @@ def setUp(self):
...
 
-        for i in range(4, 8):
-            addrs = self.nodes[i].get_addrs()
-            for addr in addrs:
-                if addr[0:4] != 'fe80':
-                    self.assertTrue(self.nodes[SED1].ping(addr))
+        # verify DUT_ROUTER1 didn't generate an Address Query Request.
+        dut_messages = self.sniffer.get_messages_sent_by(DUT_ROUTER1)
+        msg = dut_messages.next_coap_message('0.02', '/a/aq', False)
+        assert msg is None, "Error: The DUT sent an Address Query Request"
```
Comment (librasungirl) [inline tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py]: should be in above `for` block
Suggestion:
> Error: The DUT sent an unexpected Address Query Request.

**Comment by hjian2017** [comment]
@librasungirl

File: tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py [hunk 8]
```diff
@@ -50,90 +52,83 @@ def setUp(self):
...
         del self.nodes
 
     def test(self):
+        # 1
         self.nodes[LEADER].start()
         self.nodes[LEADER].set_state('leader')
         self.assertEqual(self.nodes[LEADER].get_state(), 'leader')
 
-        self.nodes[ROUTER1].start()
+        for i in range(2,8):
```
Comment (librasungirl) [inline tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py]: replace the magic number to enumeration for better readability.

File: tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py (hunk 1 above)
Comment (librasungirl) [reply tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py]: still not updated...

File: tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py [hunk 9]
```diff
@@ -50,90 +52,83 @@ def setUp(self):
...
-        self.assertEqual(self.nodes[ED4].get_state(), 'child')
+        # 2
+        for ED in [ED1, ED2, ED3, ED4]:
+            ed_mleid = self.nodes[ED].get_ip6_address(config.ADDRESS_TYPE.ML_EID)
+            self.nodes[SED1].ping(ed_mleid)
+
+        # verify DUT_ROUTER1 generated an Address Query Request to find each node's RLOC.
+        dut_messages = self.sniffer.get_messages_sent_by(DUT_ROUTER1)
+        msg = dut_messages.next_coap_message('0.02', '/a/aq')
+        command.check_address_query(msg, self.nodes[DUT_ROUTER1], config.REALM_LOCAL_ALL_ROUTERS_ADDRESS)
```
Comment (librasungirl) [inline tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py]: still not updated, should align with above `for` body

File: tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py [hunk 10]
```diff
@@ -50,90 +52,83 @@ def setUp(self):
...
+        dut_messages = self.sniffer.get_messages_sent_by(DUT_ROUTER1)
+
+        for ED in [ED1, ED2, ED3, ED4]:
+            ed_mleid = self.nodes[ED].get_ip6_address(config.ADDRESS_TYPE.ML_EID)
+            self.assertTrue(self.nodes[SED1].ping(ed_mleid))
+
+            # verify DUT_ROUTER1 didn't generate an Address Query Request.
+            dut_messages = self.sniffer.get_messages_sent_by(DUT_ROUTER1)
+            msg = dut_messages.next_coap_message('0.02', '/a/aq', False)
+            assert msg is None, "Error: The DUT sent an Address Query Request"
```
Comment (librasungirl) [inline tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py]: add ping reply check as step 4 requires.

**Comment by librasungirl** [comment]
looks like spinel-cli doesn't support `pollperiod` command. The other option is to set the `childtimeout`.

**Comment by hjian2017** [comment]
@librasungirl

**Comment by hjian2017** [comment]
@librasungirl

File: tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py [hunk 11]
```diff
@@ -50,90 +52,87 @@ def setUp(self):
...
-        time.sleep(5)
-        self.assertEqual(self.nodes[ED3].get_state(), 'child')
+        # This method flushes the message queue so calling this method again will return only the newly logged messages.
+        dut_messages = self.sniffer.get_messages_sent_by(DUT_ROUTER1)
 
-        self.nodes[ED4].start()
-        time.sleep(5)
-        self.assertEqual(self.nodes[ED4].get_state(), 'child')
+        # 2
+        for ED in [ED1, ED2, ED3, ED4]:
```
Comment (librasungirl) [inline tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py]: @hjian2017 , you may try to flush messages of DUT_ROUTER1 before each ping to ensure no legacy messages.

Comment (hjian2017) [reply tests/scripts/thread-cert/Cert_5_3_04_AddressMapCache.py]: I think it have flushed messages when use the method "get_messages_sent_by".

**Comment by hjian2017** [comment]
@librasungirl

**Pull request merged**